lib/dry/container/mocking.rb

```diff
@@ -0,0 +1,78 @@
+module Dry
+  class Container
+    module Mocking
+      class Resolver < Container::Resolver
+        def initialize(resolver)
+          @resolver = resolver
+        end
+
+        def mocks
+          @mocks ||= {}
+        end
+
+        def call(original, key)
+          container = if mocks.key?(key.to_s)
+                        mocks
+                      else
+                        original
+                      end
+
+          @resolver.call(container, key)
+        end
+
+        def key?(container, key)
+          @resolver.key?(container, key)
+        end
+
+        def mock(registry, mocks, val = nil)
+          mocks = { mocks => val } unless mocks.is_a?(Hash)
+
+          mocks.each do |key, value|
+            registry.call(self.mocks, key, value, {})
+          end
+
+          mocks.keys
+        end
+
+        def unmock(keys)
+          keys = mocks.keys if keys.empty?
+
+          keys.map(&:to_s).each do |key|
+            mocks.delete(key)
+          end
+        end
+
+        def self.wrap(resolver)
+          case resolver
+          when self
+            resolver
+          else
+            new(resolver)
+          end
+        end
+      end
+    end
+
+    module Mixin
+      def mock(mocks, val = nil)
+        config.resolver = Mocking::Resolver.wrap(config.resolver)
```

Comment:
LoL, now we are getting into [deeper waters](http://martinfowler.com/articles/mocksArentStubs.html) :stuck_out_tongue_closed_eyes: 

I like @solnic's api, but it looks too RSpec specific, thou it could be added if requested
```ruby
require "dry/container/rspec" # ?
``` 

This would leave an API like:
```ruby
# spec_helper.rb
require 'dry/container/rspec'

# To inject mocking API into container
container.enable_stubbing!

# **/*_spec.rb
container.stub(:key1)
container.unstub(:key1) # ?

# multiple keys and a block!
container.stub(:key1, :key2) do
end
```

Which could be built on top of a more generic API
Which to prevent discussions (mocks vs stubs) could use another concept, like `faking` values maybe? so it's up to you to mock, stub or just fake your keys with whatever you want on your tests.

```ruby
# spec_helper.rb
require 'dry/container/facking' # <- Name to be defined :)

# To inject mocking API into container
container.enable_facking!

# **/*_spec.rb
# fake just like register
container.fake :key, :value
container.fake :key { something }

# to unfake a single key
container.unfake :key

# to "unfake" all mocked keys
container.unfake

# to fake within a block:
container.facking key1: 'mock1', key2: 'mock2' do
  # fake key1 and key2 only while on the block
  # only key1 and key2 are cleared after
end
```

Other posible names could be `forge`, `feign`, `pretend`, `swing` :stuck_out_tongue_winking_eye: